### ISSUE TITLE
Supporting DNX Assembly Loading

### ISSUE THREAD

**Issue by steveharter** [dotnet/coreclr#2095]
Working issue to track coreclr work required to support assembly loading.

# Current Status:
Enabling Scenario 2 (intercept referenced assembly loads) in coreclr, which at that point consumers can load assemblies in a hosted manner (need to unblock that scenario). Note that scenario 1 (direct load of assembly) is currently working.

# Summary
A common mechanism should be provided so that consumer code can load assemblies suitable for hosting scenarios and in a straightforward manner without requiring #ifdefs targeting both CoreClr and desktop Clr. 

Hosting scenarios (say by project K \ dnx) require assemblies to be loaded from a given path or byte[] into a custom or the default load context. Assemblies loaded are normally found by probing specific areas including the GAC and app paths of the AppDomain, but there are hosting scenarios that require loading assemblies outside of these paths or from byte[]. 

For example, in CoreClr a hosting controller loads Assembly1 in a late-bound manner from a specific directory, but Assembly1 has references to Assembly2 which would cause Assembly2 to normally be automatically loaded by the CLR into the default load context. However, since Assembly2 is likely located in the same directory as Assembly1 it won't be found by the default loader.

The CoreClr and desktop Clr provide different mechanisms to assist in loading assemblis. For Core CLR, the abstract class AssemblyLoadContext must be subclassed and Load(AssemblyName assemblyName) overridden to call a protected helper method such as LoadFromAssemblyPath(). In addition, the instance of the subclass must be set to be the default load context via AssemblyLoadContext.InitializeDefaultContext(context). The desktop CLR provides Assembly.LoadFrom() and the AppDomain.AssemblyResolve event. Implementing this logic is cumbersome and error-prone plus these APIs are different and thus require #ifdefs if targeting both clrs.

# Consumer vs Providers
A distinction should be made between consumers and providers with respect to the various types of loaders (e.g. byte[] or path) and between specifying a context vs. obtaining the current context. 

Multiple consumers should be able to obtain the same context (the consumer does not own the context). For example, a consumer should not need to worry about specifying the context, only asking for the default one. 

This loading logic to be defined away from the consumer, and be made extensible to add new loaders. The consumer is shielded from the low-level behavior of each type of loader (path, byte[]) and uses the existing Assembly.Load() method. A consumer should not be concerned with implementing a new loader type, only specifying the loader type and specifying inputs such as assembly name or byte[].

# Scenario 1: load assembly from given path
**this code currently runs on Ubuntu**
AsmLoad.exe:
```c#
using System.Reflection;
using System.Runtime.Loader;
public class Program
{
    static void Main(string[] args)
    {
        var ctx = new MyAssemblyLoadContext();
        AssemblyLoadContext.InitializeDefaultContext(ctx);

        var asm = Assembly.Load("MyAssembly1");
    }
}
public class MyAssemblyLoadContext : AssemblyLoadContext
{
    protected override Assembly Load(AssemblyName assemblyName)
    {
        return base.LoadFromAssemblyPath("/home/steveharter/netcore/temp/" + assemblyName + ".dll");
    }
}
```
Assembly1.dll:
```c#
namespace MyAssembly1
{
    public class MyClass1
    {
    }
}
```
# Scenario 2: load referenced assemblies into default load context (on CoreClr)
Modifying Assembly1 to reference new Assembly2 does **not** run on Ubuntu:

Assembly1.dll:
```c#
namespace MyAssembly1
{
    public class MyClass1
    {
        public static void CallMe()
        {
            var c2 = new MyAssembly2.MyClass2();
        }
    }
}
```

Assembly2.dll:
```c#
namespace MyAssembly2
{
    public class MyClass2
    {
    }
}
```

sample build commands:
csc /noconfig /nostdlib /target:library /r:mscorlib.dll /delaysign /keyfile:MyKey MyAssembly2.cs
csc /noconfig /nostdlib /target:library /r:MyAssembly2.dll /r:mscorlib.dll /delaysign /keyfile:MyKey MyAssembly1.cs
csc /noconfig /nostdlib /r:mscorlib.dll /delaysign /keyfile:MyKey AsmLoad.cs

Running this does not use the AssemblyLoadContext. Instead, the final call stack during exception shows:
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean forIntrospection)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, Boolean forIntrospection)
   at System.Reflection.Assembly.Load(String assemblyString)
   at Program.Main(String[] args)

The AssemblyResolve event in full CLR can be used to address this, but that is not present in CoreClr. In addition, using Assembly.LoadFrom(assembly2) cannot be used to pre-load the referenced assembly as LoadFrom is not supported in CoreClr (see https://github.com/dotnet/coreclr/issues/1265).

# Proposal (NOTE: this is phase 1; iterate if necessary)
- From CoreClr, add any support to be able to load a given assembly (including byte[] and path) into the default load context including referenced assemblies (see the Scenario 2 above)
- Otherwise keep the existing APIs in coreClr clr the same
- No desktop clr changes (Note that only CoreClr hosting scenarios require assemblies to be loaded into the default load context.)

In a new assembly (Microsoft.AssemblyExtensions? todo) implement functionality similar to what dnx did previously, except hopefully streamlined to avoid any workarounds relating to loading into the default load context.
- Add an IAssemblyLoadContext interface as a common consumer interface for desktop clr and CoreClr
-- Specifies where assemblies will be loaded into. This is normally the default load context for CoreClr and an AppDomain for desktop clr. This is analogous to the CoreClr class AssemblyLoadContext 
- Add an IAssemblyLoader interface
-- Specifies a mechanism to load an assembly, such as from a directory
- To IAssemblyLoadContext , add a method to add an IAssemblyLoader
-- If there is more than one assembly loader, each will be asked to load the given assembly (the first one succeeds) when the regular Assembly.Load(name\byte[]) is invoked.

Implement the given interfaces for both CoreClr and desktop Clr in internal classes (using #ifdefs to target AssemblyLoadContext, AssemblyResolve event, etc). 

Also add:
- A mechanism to obtain the default IAssemblyLoadContext (see AssemblyLoadContext class and previous dnx abstractions); additional functionality can be added later to specify the default or a specific instance.

What this achieves:
1) Continued use of Assembly.Load(name\string) as primary API for normal consumers (easy, no concern over paths or assembly resolve, etc).
2) A way for the hosting code to intercept Assembly.Load() and load into the provided or default context (by enumerating the loaders and asking each)
3) An extensible way for providers to add new loader types and for consumers or hosting code to specify which loader types to use
# Older DNX artifacts
[sample DNX consumer code](https://raw.githubusercontent.com/aspnet/Entropy/dev/samples/Runtime.CustomLoader/Program.cs)
[corefx issue #4097](https://github.com/dotnet/corefx/issues/4027)
coreclr issue #1954

Dnx|CoreClr|Notes
---|---|---
[PlatformServices](https://github.com/aspnet/dnx/blob/dev/src/Microsoft.Extensions.PlatformAbstractions/PlatformServices.cs)|---|stand alone base class that exposes some accessors for the types below
[IAssemblyLoadContext](https://github.com/aspnet/dnx/blob/dev/src/Microsoft.Extensions.PlatformAbstractions/IAssemblyLoadContext.cs)|---|provides a context that assemblies load into
[IAssemblyLoader](https://github.com/aspnet/dnx/blob/dev/src/Microsoft.Extensions.PlatformAbstractions/IAssemblyLoader.cs)|---|consumers implement to add custom loading logic
[IAssemblyLoaderContainer](https://github.com/aspnet/dnx/blob/dev/src/Microsoft.Extensions.PlatformAbstractions/IAssemblyLoaderContainer.cs)|---|contains assembly loaders
[IAssemblyLoadContextAccessor](https://github.com/aspnet/dnx/blob/dev/src/Microsoft.Extensions.PlatformAbstractions/IAssemblyLoadContextAccessor.cs)|---|obtain load contexts

**Comment by muratg** [comment]
:watch:

**Comment by muratg** [comment]
Guys, where are we with this?  This is blocking MVC.

@joshfree @danroth27

**Comment by gkhanna79** [comment]
@muratg The basic APIs have been exposed (this does not include the redesign for API unification or AssemblyResolve support).

**Comment by muratg** [comment]
Thanks @gkhanna79.   

@pakrym FYI.

**Comment by gkhanna79** [comment]
I have gone through this proposal and here are my thoughts.

The example in "Current Issues and Complexities" is incorrect. It needs to pass AssemblyName instance in the call to Assembly.Load and the reason it fails with the FileLoadexception is because it is concatenating AssemblyName to the path instead of AssemblyName.Name with .dll suffix. AssemblyName.ToString returns a string that contains simple name, culture, version, PKT and clearly, a file by that name does not exist. The fixed code for the above is below and it works even today:

<pre><code>
using System;
using System.Reflection;
using System.Runtime.Loader;

namespace ConsoleApplication
{
    public class Program
    {
        public static void Main(string[] args)
        {
            var ctx = new MyAssemblyLoadContext();
            AssemblyLoadContext.InitializeDefaultContext(ctx);

            var asm = Assembly.Load(new AssemblyName("Lib"));
        }
    }
    public class MyAssemblyLoadContext : AssemblyLoadContext
    {
        protected override Assembly Load(AssemblyName assemblyName)
        {
            return base.LoadFromAssemblyPath("C:\\Github\\DefALC\\1\\" + assemblyName.Name+".dll");
        }
    }
}
</code></pre>

> From CoreClr, add any support to be able to load a given assembly (including byte[] and path) into the default load context (see the Current Issues and Complexities section above)

This is fixed via CoreCLR PR https://github.com/dotnet/coreclr/pull/2618 and CoreFX PR https://github.com/dotnet/corefx/issues/5344.

At this point, here is what I believe is the current status:

1) All required APIs have been exposed from AssemblyLoadContext to enable transitioning from DNX to CLI. 

2) Consolidation of API surface between Desktop and CoreCLR needs to be done. This is a nice to have and I dont believe it should be blocking for RC2/RTM and can come post it.

3) AssemblyResolve (a new API/surface) is not exposed for CoreCLR. 

@davidfowl @muratg Please let me know if anything from above is required for RTM.

FYI - @jkotas

**Comment by muratg** [comment]
@pakrym Could you look at (2) and (3) above and see if we need these?

**Comment by steveharter** [comment]
Gaurav, I’ve been able to run this on Ubuntu with the latest code now. Thanks. BTW, originally I tried with and without .dll extension and yes I meant to include the .dll in that sample code.

**Comment by steveharter** [comment]
@pakrym @gkhanna79 I've updated the contents to include Scenario 2, which is the ability to load referenced assemblies into the default load context. Currently this does not work with CoreClr. Will this require the AssemblyResolve event or adding support for Assembly.LoadFrom? Ideally this would work with AssemblyLoadContext

**Comment by davidfowl** [comment]
> Will this require the AssemblyResolve event or adding support for Assembly.LoadFrom? Ideally this would work with AssemblyLoadContext

Correct! This is key. 3rd parties need to be able to respond to Load without owning the LoadContext. Thats the plugin scenario.

**Comment by gkhanna79** [comment]
@steveharter Where is the referenced assembly located in Scenario 2 above? Is it not in the same location as app since that should just work?

Also, can you please fix the code for scenario 1?

> 3rd parties need to be able to respond to Load without owning the LoadContext. Thats the plugin scenario. 

This the case of 3rd parties trying to load assemblies that are not in the same location as the app. Is that correct understanding @davidfowl ?

**Comment by gkhanna79** [comment]
Keeping this open until PRs get into RC2 branch

**Comment by steveharter** [comment]
@gkhanna79 is it possible to use the existing AssemblyLoadContext instead of adding the AssemblyResolve event to intercept referenced assembly loads? That would be ideal IMO (one mechanism instead of two)

Assuming AssemblyResolve event handler, will that be expected to obtain or create a  AssemblyLoadContext in order to call one of the methods to load the assembly such as LoadFromAssemblyPath, or are we going to also support the Assembly.LoadFrom methods as typically would be used with AssemblyResolve?

**Comment by gkhanna79** [comment]
Keeping this open until the Port PRs go through.

**Comment by gkhanna79** [comment]
@steveharter Referenced assembly loads will still be intercepted via the ALC implementation's Load override. However, if Load does not find it (or cannot find as is the case for default load context ALC instance), the Resolving event will be used to give an oppurtunity to provide the assembly.

Resolving event gets the reference to the LoadContext for which it was invoked and the consumer can use it to perform the loadfromassembly/ni/stream path.

**Comment by steveharter** [comment]
Reopening temporarily to update the status and code samples

**Comment by steveharter** [comment]
I have verified that both scenarios 1 (direct load of assembly) and 2 (indirect load of assembly by coreclr through an assembly-to-assembly reference) now work with the latest code using AssemblyLoadContext. There was no need to use the AssemblyResolve event in my scenarios. This means that the dnx\project work is unblocked and can now add appropriate constructs on top of this functionality if desired.
